bluesky/magics.py

```diff
@@ -110,53 +113,161 @@ def ct(self, line):
+                low_limit = high_limit = exc.__class__.__name__
             else:
-                value = v.__class__.__name__  # e.g. 'DisconnectedError'
-                low_limit = high_limit = offset = ''
+                low_limit = np.round(low_limit, decimals=prec)
+                high_limit = np.round(high_limit, decimals=prec)
+            try:
+                offset = p.user_offset.get()
+            except Exception as exc:
+                offset = exc.__class__.__name__
+            else:
+                offset = np.round(offset, decimals=prec)
+        else:
+            value = v.__class__.__name__  # e.g. 'DisconnectedError'
+            low_limit = high_limit = offset = ''
+
+        lines.append(LINE_FMT.format(p.name, value, low_limit, high_limit,
+                                     offset))
+    print('\n'.join(lines))
+
+
+def _labeled_devices(user_ns=None, maxdepth=6):
+    ''' Returns list of all devices that are labeled.
+
+        Parameters
+        ----------
+        user_ns : dict, optional
+            The namespace to search on
+
+        maxdepth: int, optional
+            max recursion depth
+
+        Returns
+        -------
+            A dictionary of (name, ophydobject) tuple indexed by device label.
+
+        Examples
+        --------
+        Read devices labeled as motors:
+            objs = _labeled_devices()
+            my_motors = objs['motors']
+    '''
+    # could be set but lists are more common for users
+    obj_list = collections.defaultdict(list)
+
+    if maxdepth <= 0:
+        print("Recursion limit exceeded")
+        return obj_list
+
+    if user_ns is None:
+        user_ns = get_ipython().user_ns
+
+    for key, obj in user_ns.items():
+        # ignore objects beginning with "_"
+        # (mainly for ipython stored objs from command line
+        # return of commands)
+        # also check its a subclass of desired classes
+        if not key.startswith("_"):
+            if is_parent(obj):
+                labels = getattr(obj, '_ophyd_labels_', set())
+                obj_list.update(_labeled_devices(user_ns=obj.__dict__,
+                                              maxdepth=maxdepth-1,))
+            else:
+                if hasattr(obj, '_ophyd_labels_'):
+                    # inherit parent labels
+                    labels = obj._ophyd_labels_
+                    for label in labels:
+                        obj_list[label].append((key, obj))
+
+    return obj_list
+
 
-            lines.append(LINE_FMT.format(p.name, value, low_limit, high_limit,
-                                         offset))
-        print('\n'.join(lines))
+def is_parent(dev):
+    # return whether a node is a parent
+    # should not have component_names, or if yes, should be empty
+    # read_attrs needed to check it's an instance and not class itself
+    return (hasattr(dev, 'component_names') and len(dev.component_names) > 0
+            and hasattr(dev, 'read_attrs'))
```

Comment:
I think `not isinstance(dev, type)` is safer than `hasattr(dev, 'read_attrs')`, even though this will probably be fine.

Reply:
great idea! done